Search: add search setting page with only toggles

kangzj opened the pull request:
#### Changes proposed in this Pull Request:

Added search setting page, including Search and Instant Search toggles and a mocked instant search dialog.

#### Jetpack product discussion
pcNPJE-7J#comment-161

#### Does this pull request change what data or activity we track or use?
New tracks will be added in separate PR.

#### Testing instructions:
- Go to a site with Jetpack Instant Search subscription (enter URL directly `/wp-admin/admin.php?page=jetpack-search&a8ctest`)
- Ensure `Search` sub menu is added under Jetpack
- Click the `Search` sub menu
- Ensure the page is accessible with both PC and mobile devices
- Ensure the toggles work as expected
![image](https://user-images.githubusercontent.com/1425433/123947963-1061ec80-d9f5-11eb-8380-19848bdf2584.png)


NOTE: 
- The functionality is hidden, unless a `a8ctest` param is detected in URL, e.g. `/wp-admin/admin.php?page=jetpack-search&a8ctest`.
- `class.jetpack-admin-page.php` is moved to `projects/plugins/jetpack/_inc/lib/admin-pages/`, because if we disabled search module, Jetpack wouldn't load `modules/search` anymore. And I think it makes more sense to be there.
- Just trying to make the PR as small as possible - footer, buttons, Legacy search mock and tracking will be added in the follow-up PRs.

kangzj committed:
[not verified] add search dashboard sub menu and page

kangzj committed:
[not verified] added frontend

kangzj committed:
[not verified] stretch the toggle div to cover wp admin footer

kangzj committed:
[not verified] added flag to show the search setting sub menu when a8ctest param exists

kangzj committed:
[not verified] added should_add_sub_menu

kangzj committed:
[not verified] added mocked legacy search

kangzj committed:
[not verified] added loading screen

kangzj committed:
[not verified] changelog

kangzj committed:
[not verified] removed unnecessary code

kangzj commented on projects/plugins/jetpack/class.jetpack-admin.php:
```diff
@@ -37,8 +37,12 @@ private function __construct() {
+		$this->jetpack_search = new Jetpack_Search_Dashboard_Page();
+
 		add_action( 'admin_menu', array( $this->jetpack_react, 'add_actions' ), 998 );
 		add_action( 'admin_menu', array( $this->jetpack_react, 'add_actions' ), 998 );
```
dup lines, probably an error in existing code. will address in another PR.

kangzj replied:
will be addressed in #20258

jsnmoon commented on projects/plugins/jetpack/_inc/client/search-dashboard-entry.js:
```diff
@@ -0,0 +1,37 @@
+	);
+}
+
+// Initialize Instant Search when DOMContentLoaded is fired, or immediately if it already has been.
```
Needs a comment update; this isn't the Instant Search application.

```suggestion
// Initialize the dashboard when DOMContentLoaded is fired, or immediately if it already has been.
```

jsnmoon commented on projects/plugins/jetpack/_inc/client/search/dashboard/_variables.scss:
```diff
@@ -0,0 +1,8 @@
+@import '../../scss/mixin_breakpoint.scss';
+
+// Colors
+$color-jetpack-green: #069e08;
```
We could import this from `_inc/client/scss/variables/_colors.scss` instead of redefining it here.

kangzj replied:
Good point. addressed in 413a0f0

jsnmoon commented on projects/plugins/jetpack/_inc/client/search/dashboard/index.jsx:
```diff
@@ -0,0 +1,90 @@
+								) }
+							</h1>
+						</div>
+						<div className="jp-search-dashboard__search-dialog">
```
Nit: why "dialog"? I'm not seeing the connection between "dialog" and the mocked instant search visuals.

kangzj replied:
renamed to from `search-dialog` to `mocked-search-interface`

jsnmoon commented on projects/plugins/jetpack/_inc/client/search/dashboard/index.jsx:
```diff
@@ -0,0 +1,90 @@
+		return {
+			apiRootUrl: getApiRootUrl( state ),
+			apiNonce: getApiNonce( state ),
+			isBusinessPlan: 'is-business-plan' === planClass,
```
This seems like an unused property for this component.

kangzj replied:
removed in e4b7c72f6

jsnmoon commented on projects/plugins/jetpack/_inc/client/search/dashboard/index.jsx:
```diff
@@ -0,0 +1,90 @@
+	useComponentWillMount( () => {
+		apiRootUrl && restApi.setApiRoot( apiRootUrl );
+		apiNonce && restApi.setApiNonce( apiNonce );
+		dispatchedSetInitialState && dispatchedSetInitialState();
```
Nit: I've not seen this pattern of prefixing the variable with `dispatched`; is this a common pattern elsewhere? Otherwise, I'd recommend removing it.

kangzj replied:
The piece of code is from the performance dashboard. I guess they just wanted a different name as the selector has `setInitialState` as its name. Renamed to `setSearchDashboardInitialState` in fe9106924

jsnmoon commented on projects/plugins/jetpack/_inc/client/search/dashboard/index.jsx:
```diff
@@ -0,0 +1,90 @@
+			apiNonce: getApiNonce( state ),
+			isBusinessPlan: 'is-business-plan' === planClass,
+			isLoading: isFetchingSitePurchases( state ),
+			hasActiveSearchPurchase: selectHasActiveSearchPurchase( state ),
```
This also seems like an unused property for this component.

kangzj replied:
removed in e4b7c72f6

jsnmoon reviewed (commented):
Left some comments on code quality, but this PR works as expected.

Aside: the query string feature gate doesn't work very well, but that can be ignored since it's meant to be a temporary measure. (E.g. Start at `/wp-admin/?a8ctest` and clicking on `Jetpack -> Search` will land you on a 403 page; navigating directly to `/wp-admin/admin.php?page=jetpack-search&a8ctest` works as expected.)

Feel free to mark ready for crew review!

kangzj committed:
[not verified] Update projects/plugins/jetpack/_inc/client/search-dashboard-entry.js

Co-authored-by: Jason Moon <4044428+jsnmoon@users.noreply.github.com>

kangzj committed:
[not verified] reuse existing color definition

kangzj committed:
[not verified] renaming setSearchDashboardInitialState

kangzj committed:
[not verified] renaming dialog to interface

kangzj committed:
[not verified] removed unused variables

kangzj commented:
> Left some comments on code quality, but this PR works as expected.
> 
> Aside: the query string feature gate doesn't work very well, but that can be ignored since it's meant to be a temporary measure. (E.g. Start at `/wp-admin/?a8ctest` and clicking on `Jetpack -> Search` will land you on a 403 page; navigating directly to `/wp-admin/admin.php?page=jetpack-search&a8ctest` works as expected.)
> 
> Feel free to mark ready for crew review!

Thanks for the comments @jsnmoon 👍 
Addressed all feedback, rebased and marked ready for crew review 🎉

kangzj committed:
[not verified] smaller fonts on mobile devices

kangzj committed:
[not verified] adjust bottom sytling

kangzj commented:
Hey @bluefuton do you think we need designer's review at the stage or before we turn the page on? I'm thinking the latter. Thanks.

bluefuton commented:
> Hey @bluefuton do you think we need designer's review at the stage or before we turn the page on? I'm thinking the latter.

Before launch would work 👍 Easier to address design feedback in a followup PR I think.

jeherve commented on projects/plugins/jetpack/_inc/lib/admin-pages/class-jetpack-search-dashboard-page.php:
```diff
@@ -0,0 +1,141 @@
+	protected function should_add_sub_menu() {
+		// TODO: temporary flag for testing. Will be removed on the last PR merge.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		return Jetpack_Plan::supports( 'search' ) && array_key_exists( 'a8ctest', $_GET );
```
Should we check if the site is connected / not in offline mode here?

jsnmoon replied:
If the site isn't connected or is offline, I see two paths forward:

- Don't add the search dashboard submenu link. This falls back on the existing connection UX built into Jetpack.
- Add the submenu link, but use the dashboard to direct the user to the connection process. (Or to bring them out of offline mode? I'm not very familiar with offline mode.)

Context: We'd like to eventually make this the landing page when Jetpack Search becomes its own plugin via RNA.

kangzj replied:
> Should we check if the site is connected / not in offline mode here?

This class extends `Jetpack_Admin_Page` which does the test before calling the function to add the submenu.
https://github.com/Automattic/jetpack/blob/367da565388df030e9c2ef760cfdb916b912f5ff/projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-admin-page.php#L69

kangzj replied:
> If the site isn't connected or is offline, I see two paths forward:
> 
>     * Don't add the search dashboard submenu link. This falls back on the existing connection UX built into Jetpack.
> 
>     * Add the submenu link, but use the dashboard to direct the user to the connection process. (Or to bring them out of offline mode? I'm not very familiar with offline mode.)
> 
> 
> Context: We'd like to eventually make this the landing page when Jetpack Search becomes its own plugin via RNA.

It currently does the first, but we could def talk about the second.

jeherve replied:
> This class extends `Jetpack_Admin_Page` which does the test before calling the function to add the submenu.

That's perfect for now I think 👍

jeherve commented on projects/plugins/jetpack/_inc/client/search/dashboard/index.jsx:
```diff
@@ -0,0 +1,82 @@
+						<div className="jp-search-dashboard__title">
+							<h1>
+								{ __(
+									"Help your visitors find exactly what they're looking for, fast",
```
Should we have a period here?

kangzj replied:
As per the design there's not. @keoshi might be the right one to answer this :-)

kangzj replied:
@keoshi is currently on leave and we'll ask him to have a finally designer's review before we turn on the switch, so we'll stick to the design for now. Thanks.

jeherve reviewed (commented):
This looks really good! ✨ 

I have a few questions below. I'll let you decide what's a blocker, what's not, I don't know if you've already planned some of those things for a v2?

******

I think it would make sense to add the footer we use on the other Jetpack dashboard pages, to be consistant:

![image](https://user-images.githubusercontent.com/426388/124301786-964e7680-db60-11eb-8dba-b87802ee7741.png)

******

Similarly, should we have the common navigation items we have on the other dashboard pages?

![image](https://user-images.githubusercontent.com/426388/124302229-35736e00-db61-11eb-99f8-7b8c65f7115f.png)

*******

There seems to be a small issue with margins on the right of the search container on mobile:

![Screenshot_20210702-181558](https://user-images.githubusercontent.com/426388/124302633-c0546880-db61-11eb-9f08-8cc777e59b29.jpg)

********

There may also be some issues on height-limited screens:

![image](https://user-images.githubusercontent.com/426388/124302838-fd205f80-db61-11eb-9332-5d746748644d.png)

*********

It would be nice if there were notices when toggle status changes, like in the rest of the dashboard.

jsnmoon commented:
> Similarly, should we have the common navigation items we have on the other dashboard pages? (@jeherve)
> 
> ![image](https://user-images.githubusercontent.com/426388/124302229-35736e00-db61-11eb-99f8-7b8c65f7115f.png)

I'll push back slightly on this, given the design specs at p1HpG7-cdF-p2. We'll eventually want to implement a new aesthetic for tabbed navigation, like so:

<img width="1138" alt="Screen Shot 2021-07-02 at 5 19 41 PM" src="https://user-images.githubusercontent.com/4044428/124336029-b29ae500-db59-11eb-861d-615fcaf63464.png">

kangzj committed:
[not verified] use components/gridicon in client

kangzj commented on projects/plugins/jetpack/_inc/client/search/dashboard/mocked-instant-search.jsx:
```diff
@@ -0,0 +1,79 @@
+ * Internal dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import Gridicon from '../../../../modules/search/instant-search/components/gridicon';
```
thanks. addressed in 992c4cbfc

kangzj committed:
[not verified] added key to looped item

kangzj commented on projects/plugins/jetpack/_inc/client/search/dashboard/mocked-instant-search.jsx:
```diff
@@ -0,0 +1,79 @@
+ */
+export default function MockedInstantSearch() {
+	const renderFilterOption = label => (
+		<div className="jp-mocked-instant-search__search-filter">
+			<label>
+				<input type="checkbox" disabled="disabled" />{ ' ' }
+				<TextRowPlaceHolder style={ { width: '30%' } } />
+			</label>
+		</div>
+	);
+	const renderSearchResult = () => (
+		<div className="jp-mocked-instant-search__search-result">
+			<TextRowPlaceHolder
+				style={ { height: '2.5em', width: '50%', margin: '0.1em 0.1em 1em 0.1em' } }
+			/>
+			<TextRowPlaceHolder style={ { height: '1em', width: '90%', margin: '0.1em' } } />
+			<TextRowPlaceHolder style={ { height: '1em', width: '70%', margin: '0.1em' } } />
+		</div>
```
Good catch. These item literally will only be rendered once, so I didn't add keys. But definite a good thing to be consistent. addressed in a434f2882.

kangzj committed:
[not verified] added translation for Filter Options

kangzj commented on projects/plugins/jetpack/_inc/client/search/dashboard/mocked-instant-search.jsx:
```diff
@@ -0,0 +1,79 @@
+					</div>
+				</div>
+				<div className="jp-mocked-instant-search__search-results-secondary">
+					<div className="jp-mocked-instant-search__search-filter-header">Filter Options</div>
```
thanks you've got keen eyes :-)
addressed in 73253de84

kangzj committed:
[not verified] fixed margin issue for mobile device

kangzj committed:
[not verified] fixed style for height limited device

kangzj commented:
Hi @jeherve thank you very much for all the comments 👍 

> I think it would make sense to add the footer we use on the other Jetpack dashboard pages, to be consistant:

Footer will be added in [another PR](https://github.com/Automattic/jetpack/pull/20252), just wanted to make the first PR as small as possible.

> Similarly, should we have the common navigation items we have on the other dashboard pages?

As @jsnmoon mentioned, this is designed but we decided not to include it in this PR.

> There seems to be a small issue with margins on the right of the search container on mobile:

Fixed in cd9506949

> There may also be some issues on height-limited screens:

Fixed in 2628891a2

kangzj committed:
[not verified] optimized responsitive design

kangzj committed:
[not verified] optimized responsitive design

kangzj committed:
[not verified] move height control to outer element

bluefuton commented:
Just wanted to flag an accessibility issue - it doesn't seem to be able to access the form toggles using the keyboard only. This might be a problem in the underlying components?

<img width="1047" alt="Screen Shot 2021-07-06 at 14 20 57" src="https://user-images.githubusercontent.com/17325/124533030-bbdfc680-de65-11eb-8a22-1470795ded26.png">

<img width="866" alt="Screen Shot 2021-07-06 at 14 24 34" src="https://user-images.githubusercontent.com/17325/124533177-f5183680-de65-11eb-8eaf-41ca6ac989e3.png">

bluefuton commented:
The explanatory text looks different in the designs from @keoshi - not italic, and text is darker.

This PR:

<img width="835" alt="Screen Shot 2021-07-06 at 14 32 10" src="https://user-images.githubusercontent.com/17325/124533815-23e2dc80-de67-11eb-8c82-ac66dbd1762f.png">

Designs i3 (p1HpG7-cdF-p2):

<img width="554" alt="Screen Shot 2021-07-06 at 14 31 55" src="https://user-images.githubusercontent.com/17325/124533829-2a715400-de67-11eb-94fd-bd59ca96bb95.png">

kangzj committed:
[not verified] add style for toggle focus state

kangzj committed:
[not verified] explanation style

kangzj committed:
[not verified] adjust bottom div height

kangzj committed:
[not verified] changed explanation font smaller

kangzj commented:
> Just wanted to flag an accessibility issue - it doesn't seem to be able to access the form toggles using the keyboard only. This might be a problem in the underlying components?
> 

Good catch. Fixed in 6070fe9
![watch](https://user-images.githubusercontent.com/1425433/124544976-4e8b6000-de7c-11eb-99e8-2e69c1d315a6.gif)

>The explanatory text looks different in the designs from @keoshi - not italic, and text is darker.

Fixed in cf06779

bluefuton commented on projects/plugins/jetpack/_inc/client/search/dashboard/module-control.jsx:
```diff
@@ -0,0 +1,153 @@
+								<p className="jp-form-search-setting__explanation">{ SEARCH_DESCRIPTION }</p>
+							</div>
+						</div>
+						<div className="jp-search-search-toggle jp-search-search-toggle--instant-search">
```
We've got a mixture of underscore and dash separators in here - `jp-search-search-toggle__description` vs `jp-search-search-toggle--instant-search` - can we use underscores consistently after `jp-search-search-toggle`?

kangzj replied:
`jp-search-search-toggle--instant-search` is meant to be a block modifier while `jp-form-search-setting__explanation` is an element, which is the reason that the naming is different according to [BEM](http://getbem.com/naming/). Does that make sense?

bluefuton replied:
Thanks for explaining @kangzj! BEM seems like a good idea but we should clarify what the right CSS standards to follow here are.

I think generally speaking we've used the Calypso CSS guidelines for the Instant Search app itself:

https://github.com/Automattic/wp-calypso/blob/trunk/docs/coding-guidelines/css.md

...and Jetpack just follows the WordPress Core CSS guidelines by the look of the docs:

https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/

Is that correct @jeherve?

jeherve commented:
> We'll eventually want to implement a new aesthetic for tabbed navigation, like so:

This will look good, that will be super nice! Can't wait until we have the same design for the rest of the Jetpack dashboard. :)

kangzj committed:
[not verified] make css class name consistent

kangzj committed:
added missing hook

kangzj commented:
Hi @jeherve good day 🌞 
Would you mind having another look at this? Thanks heaps. 👍

jeherve reviewed (approved):
This looks good to me right now. In future PRs I think there are 2 more things that would be nice to address, but it may already be on your to-do list :)

1. I think it'd be nice if we didn't have the main module toggle switching status when I only want to disable Instant Search:

https://user-images.githubusercontent.com/426388/124637907-9fa25080-de8a-11eb-9707-ed6091bba4d9.mov

2. Down the road I don't think we should have to refresh the page to switch between the Search dashboard and the other Jetpack dashboard pages.

kangzj commented:
Thanks that'll allow me to further polish the page 👍 

>      I think it'd be nice if we didn't have the main module toggle switching status when I only want to disable Instant Search:

Do you mean the Search toggle should be available while toggling Instant Search? The experience is aligned with the existing Search toggling on Performance page. Are you sure you want me to do so?


>     Down the road I don't think we should have to refresh the page to switch between the Search dashboard and the other Jetpack dashboard pages.

Exactly. I think the whole experience is designed and on the way `p6TEKc-4jU`.

bluefuton commented:
One final query from me on this one: the mocked search UI works well, but for simplicity's sake, could we just use some responsive images instead?

jeherve commented:
> Do you mean the Search toggle should be available while toggling Instant Search? The experience is aligned with the existing Search toggling on Performance page. Are you sure you want me to do so?

No you're right, that's probably safer that way. Having the two toggles "change" was a bit confusing, but it's probably better that way. You can ignore me :)

jsnmoon commented:
> One final query from me on this one: the mocked search UI works well, but for simplicity's sake, could we just use some responsive images instead?

[This](https://css-tricks.com/responsive-images-css/) might be a useful resource for responsive images.
